Validate bind group layouts by value

yanchith opened the pull request:
This PR widens the comparisons done between two bind group layouts to also consider them equivalent if their vectors of `BindGroupLayoutBinding`s are equal. 

Tested on various modifications of the cube example from wgpu-rs. https://gist.github.com/yanchith/56e15fe35658b14587ff9bfcbd53116f

Fixes #335

kvark commented on wgpu-core/src/command/bind.rs:
```diff
@@ -98,29 +98,56 @@ impl BindGroupEntry {
-                }) if layout_id == bind_group_layout_id => {
-                    LayoutChange::Match(group_id.value, &self.dynamic_offsets)
+                }) => {
+                    if layout_id == bind_group_layout_id {
```
I think we can do even better than this. What if we resolve the duplicate layouts earlier, in `create_bind_group_layout`? If we see that a layout with the same content already exists, we can just return the same ID. This means we'd always be able to compare by ID.

yanchith replied:
Nice! I'll give it a shot in the coming days 👍

kvark replied:
Thanks! And sorry for coming up with this late. I had a feeling that we were missing something, but didn't realize what it was until now :sweat_smile:

yanchith commented:
Alright, the changes you requested have some really nice properties: all the work is done at init time and the code is much simpler too :+1: 

Couple of caveats:

- The bind group layout storage is currently scanned linearly to see if any bind group layouts have the same bindings. Not sure how expensive the call is considered to be, but we could use an additional or better suited data structure to make the search faster.  
- I added the `Backend` to the storage's value type, because I had to produce it for the ID. Let's hope it doesn't pessimize the data layout much.
- Now that the backend is available in the storage, I added a couple of asserts here and there: `assert_eq!(backend, storage_backend)`, similar as was already done for epochs. Not sure if wgpu currently supports running on multiple backends simultaneously, so maybe these new asserts are totally redundant :joy: 
- If there ever is a `wgpu_bind_group_layout_destroy` (not sure if it is planned), we will need to rething this approach so we don't drop the layout while still in use.

kvark commented on wgpu-core/src/device.rs:
```diff
@@ -1182,6 +1182,15 @@ impl<F: IdentityFilter<BindGroupLayoutId>> Global<F> {
 
+        {
+            let (bind_group_layout_guard, _) = hub.bind_group_layouts.read(&mut token);
+            for (id, value) in bind_group_layout_guard.iter() {
```
let's use `bind_group_layout_guard.iter().any(|| {..})`

kvark commented on wgpu-core/src/hub.rs:
```diff
@@ -89,55 +89,67 @@ impl IdentityManager {
 pub struct Storage<T, I: TypedId> {
     //TODO: consider concurrent hashmap?
-    map: VecMap<(T, Epoch)>,
+    map: VecMap<(T, Epoch, Backend)>,
```
this is unexpected, why do we need to add the backend here?
Each storage, conceptually, only stores things for the same backend. There can't be two different indices of that map with different backends.

yanchith replied:
Okay, I thought as much, but good to know! I will have to figure out a way to produce the `TypedId` for the iterator (need a backend to `zip` the `TypedId`), but maybe I can just pass the backend to `iter()`...

kvark commented:
You are right that we'll have an issue with deleting the bind groups, and that's something we need to have sooner rather than later.

kvark commented:
bors r-
sorry, just something I clarified only now

kvark commented:
@yanchith can we please store the bindings in a map instead of a vec (i.e. binding number -> value)? we want to make sure the contents of the bindings match by value regardless of the order they are specified.
Also, please squash the commits.
Sorry about getting late to you!

yanchith commented:
No prob, I'll get to it tomorrow :+1:

kvark commented:
Everything that can be deleted is supposed to be refcounted internally, including the bind group layouts.
I think what we'll be doing is just manually bumping the refcount when the ID is re-used, and decreasing it when the user requests to delete it.

yanchith commented:
Done as you requested. There is a caveat now, documented in a TODO I added. We can easily avoid it at the cost of allocating a hashmap that we possibly might throw away right after.

yanchith commented on wgpu-core/src/device.rs:
```diff
@@ -1203,7 +1230,7 @@ impl<F: IdentityFilter<BindGroupLayoutId>> Global<F> {
         let layout = binding_model::BindGroupLayout {
             raw,
-            bindings: bindings.to_vec(),
+            bindings: bindings.iter().cloned().map(|b| (b.binding, b)).collect(),
```
This is the allocation that we can move up to simplify the comparison and also make it more correct. But we may end up not using it.

kvark commented on wgpu-core/src/device.rs:
```diff
@@ -1182,6 +1182,34 @@ impl<F: IdentityFilter<BindGroupLayoutId>> Global<F> {
+                bind_group_layout_guard
+                    .iter(device_id.backend())
+                    .find(|(_, bgl)| {
+                        bindings.len() == bgl.bindings.len()
```
ok, I see what you did here. Let's move the hashmap creation early and just compare hashmaps here directly. The code would be simpler, and bind group layout creation is not a place that needs optimization - throwing away the hashmap is fine.

kvark commented:
Sorry it took us long, and congrats on your contribution!

yanchith commented:
No prob, happy to help again in the future!
